firedrake/slate/slate.py

```diff
@@ -0,0 +1,581 @@
+       node class; is not meant to be worked with directly. Only use
+       the appropriate subclasses.
+    """
+
+    id = 0
+
+    def __init__(self):
+        """Constructor for the TensorBase abstract class."""
+        self.id = TensorBase.id
+        TensorBase.id += 1
+        self._hash = None
+
+    @cached_property
+    def shapes(self):
+        """Computes the internal shape information of its components.
+        This is particularly useful to know if the tensor comes from a
+        mixed form.
+        """
+        shapes = {}
+        for i, arg in enumerate(self.arguments()):
+            V = arg.function_space()
+            shapelist = []
+            for fs in V:
+                shapelist.append(fs.fiat_element.space_dimension() * fs.dim)
+            shapes[i] = tuple(shapelist)
+        return shapes
+
+    @cached_property
+    def shape(self):
+        """Computes the shape information of the local tensor."""
+        shape = []
+        for shapelist in self.shapes.values():
+            shape.append(sum(shapelist))
+        return tuple(shape)
+
+    @cached_property
+    def rank(self):
+        """Returns the rank information of the tensor object."""
+        return len(self.arguments())
+
+    @property
+    def inv(self):
+        return Inverse(self)
+
+    @property
+    def T(self):
+        return Transpose(self)
+
+    def __add__(self, other):
+        return TensorAdd(self, other)
+
+    def __radd__(self, other):
+        """Ordering of tensor addition does not matter."""
+        return self.__add__(other)
+
+    def __sub__(self, other):
+        return TensorSub(self, other)
+
+    def __rsub__(self, other):
+        """Ordering of tensor subtraction does not matter."""
+        return self.__sub__(other)
+
+    def __mul__(self, other):
+        # if other is a ufl.Coefficient, return action
+        if isinstance(other, Coefficient):
+            return TensorAction(self, other)
+        return TensorMul(self, other)
+
+    def __rmul__(self, other):
+        """Tensor multiplication is not commutative in general."""
+        assert isinstance(other, TensorBase)
+        return other.__mul__(self)
+
+    def __neg__(self):
+        return Negative(self)
+
+    def ufl_domain(self):
+        """This function returns a single domain of integration occuring
+        in the tensor.
+
+        The function will fail if multiple domains are found.
+        """
+        domains = self.ufl_domains()
+        assert all(domain == domains[0] for domain in domains), "All integrals must share the same domain of integration."
+
+        return domains[0]
+
+    def __hash__(self):
+        """Returns a hash code for use in dictionary objects."""
+        if self._hash is None:
+            self._hash = hash((type(self), )
+                              + tuple(hash(a) for a in self.arguments())
+                              + tuple(hash(c) for c in self.coefficients()))
+        return self._hash
```

Comment:
Sounds good. I won't be able in tomorrow due to CDT stuff, but I will be in on Thursday. In the meantime, I'll read up on hashes.